### PR TITLE
add cached-path file utils

### LINKED ISSUE #72: Consider using cached-path crate in resources module
Hey @guillaume-be, I wrote a Rust version of the functionality you find in `file_utils.py` from `transformers` (and `allennlp`) called [`cached-path`](https://github.com/epwalsh/rust-cached-path). This could make the resources module more robust. `cached-path` knows when to update a resource based on the Etag, and ensures the cache is never corrupted if a download fails for any reason. It also avoids race conditions by using file locks.

Let me know what you think.

### PR DESCRIPTION
**Pull request by epwalsh**
closes #72 

So far I've just added a `common/file_utils.rs` module. The idea being that this would replace what's currently in `common/resources.rs`.

My question for you is, do you want to keep the `LocalResource` and `RemoteResource` abstractions?

### CODE REVIEW

**Comment by guillaume-be** [comment]
@epwalsh thank you for the suggestions. I believe the `cached_path` functionality could be easily added without causing breaking changes. My idea was to just replace the body of the existing https://github.com/guillaume-be/rust-bert/blob/a52279e24d879db2a4b79833ac409cecd6b3f2ff/src/common/resources.rs#L186 and replace the check for the existing file & download by a call to `cached_path`. Could you please use the existing file and methods in `resources` instead of creating a separate `file_utils` file?

For now I'd like to keep the concept of local and remote resources - and I believe this would still work with the proposed changes.

File: Cargo.toml [hunk 1]
```diff
@@ -37,6 +37,8 @@ serde = { version = "1.0.114", features = ["derive"] }
 dirs = "3.0.1"
 itertools = "0.9.0"
 ordered-float = "2.0.0"
+cached-path = "0.4.0-rc1"
```
Comment (guillaume-be) [inline Cargo.toml:40]: could you please ensure a release is added as a dependency for the final version of this PR?

File: src/common/file_utils.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+use crate::common::error::RustBertError;
```
Comment (guillaume-be) [inline src/common/file_utils.rs:1]: Please use the existing `resources` file for these utilities

File: src/common/file_utils.rs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+use std::path::PathBuf;
+
+lazy_static! {
+    pub static ref CACHE: Cache = Cache::builder()
+        .dir(_get_cache_directory())
+        .build()
+        .unwrap();
+}
+
+fn _get_cache_directory() -> PathBuf {
```
Comment (guillaume-be) [inline src/common/file_utils.rs:14]: This already exists at https://github.com/guillaume-be/rust-bert/blob/a52279e24d879db2a4b79833ac409cecd6b3f2ff/src/common/resources.rs#L150

File: src/common/file_utils.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            let mut home = dirs::home_dir().unwrap();
+            home.push(".cache");
+            home.push(".rustbert");
+            home
+        }
+    };
+    home
+}
+
+pub fn cached_path(resource: &str) -> Result<PathBuf, RustBertError> {
```
Comment (guillaume-be) [inline src/common/file_utils.rs:27]: Could you please wrap the cached path functionality directly in https://github.com/guillaume-be/rust-bert/blob/a52279e24d879db2a4b79833ac409cecd6b3f2ff/src/common/resources.rs#L186 ?

**Review by guillaume-be — CHANGES_REQUESTED** [review]

**Comment by epwalsh** [comment]
Hey @guillaume-be, no problem.

One issue though is that, like `transformers/file_utils.py`, `cached-path` automatically generates a name for the files it downloads. The name is generated from a hash of the resource string and the ETag. But your `RemoteResource` struct takes a `target_path`.

I think I could make `cached-path` work with custom target paths but I'll have to change some things.

**Comment by guillaume-be** [comment]
What if the target_path would point to a directory instead (i.e. remove the file name). 
```rust
impl BartModelResources {
    /// Shared under MIT license by the Facebook AI Research Fairseq team at https://github.com/pytorch/fairseq. Modified with conversion to C-array format.
    pub const BART: (&'static str, &'static str) = (
        "bart/model.ot",
        "https://cdn.huggingface.co/facebook/bart-large/rust_model.ot",
    );
}
```
Would for example become:
```rust
impl BartModelResources {
    /// Shared under MIT license by the Facebook AI Research Fairseq team at https://github.com/pytorch/fairseq. Modified with conversion to C-array format.
    pub const BART: (&'static str, &'static str) = (
        "bart",
        "https://cdn.huggingface.co/facebook/bart-large/rust_model.ot",
    );
}
```
and the `.ot` file would be dynamically generated by `cached_path`?

**Comment by epwalsh** [comment]
That seems reasonable

**Comment by epwalsh** [comment]
Just made some updates. CI will probably fail. Let me know if you're happy with the API and then I'll fix the compile errors. IMHO though it doesn't make sense to have a `download_resource` function when we get the same functionality from `Resource.get_local_path()`.

Let me know.

File: src/common/resources.rs [hunk 5]
```diff
@@ -39,7 +36,10 @@ pub enum Resource {
 }
 
 impl Resource {
-    /// Gets the local path for a given resource
+    /// Gets the local path for a given resource.
+    ///
+    /// If the resource is a remote resource, it is downlaoded and cached. Then the path
```
Comment (guillaume-be) [inline src/common/resources.rs:41]: typo

**Comment by guillaume-be** [comment]
@epwalsh Thank you for the change. The `get_local_path` indeed is more idiomatic and is a nicer way to get to access local resources or download remote files. I wanted to avoid breaking changes - maybe the `download_function` could be kept in parallel to the now standard `get_local_path` method, but deprecated for a version cycle, i.e.
```
#[deprecated(
    since = "0.9.1",
    note = "Please use `Resource.get_locao_path()` instead"
)]
```

**Comment by guillaume-be** [comment]
@epwalsh Note that I am unable to build the `cached-path` crate on my machine (Windows). The `file-lock` dependency seems to be incompatible with my setup. This may be because `gcc` is listed as a dependency for this crate while the MSVC is used as a standard toolchain for windows machine. I have added a Windows build to the CI, hopefully it catches the issue.

**Comment by epwalsh** [comment]
I'll add a Windows build to `cached-path` as well as see if I can fix the issue.

**Comment by epwalsh** [comment]
Okay, should build on Windows now 👌

**Comment by epwalsh** [comment]
Still TODO:
- [ ] Update resources like https://github.com/guillaume-be/rust-bert/pull/74#issuecomment-690615106
- [ ] Remove use of `download_resource` from examples and docs

**Comment by guillaume-be** [comment]
@epwalsh This looks great thanks. 

Actually regarding the first TODO - one could potentially leave it the way it is. As you can see in the CI, one nice effect is that the target file name becomes the name of the folder where the cached version will be stored. This is quite nice as the hashed version makes it difficult to differentiate a `config.json` from a `vocab.txt` otherwise. The current path makes it easy to know exactly what each cache file corresponds to.

What do you think?

**Comment by epwalsh** [comment]
> @epwalsh This looks great thanks.
> 
> Actually regarding the first TODO - one could potentially leave it the way it is. As you can see in the CI, one nice effect is that the target file name becomes the name of the folder where the cached version will be stored. This is quite nice as the hashed version makes it difficult to differentiate a `config.json` from a `vocab.txt` otherwise. The current path makes it easy to know exactly what each cache file corresponds to.
> 
> What do you think?

I'm fine with that if you are 👍

File: src/common/resources.rs [hunk 6]
```diff
@@ -176,37 +185,13 @@ fn _get_cache_directory() -> PathBuf {
...
-                    output_file.flush().await?;
-                    Ok::<(), RustBertError>(())
-                })?;
-            }
-            Ok(resource.get_local_path())
-        }
-        Resource::Local(_) => Ok(resource.get_local_path()),
-    }
+pub fn download_resource(resource: &Resource) -> Result<PathBuf, RustBertError> {
+    Ok(resource.get_local_path()?)
```
Comment (guillaume-be) [inline src/common/resources.rs:196]: just nitpicking, but I believe you could return `resource.get_local_path()` directly as this is a `Result<PathBuf, RustBertError>`

**Comment by epwalsh** [comment]
@guillaume-be I noticed there were bunch of warnings (unrelated to this PR) when I ran clippy:

```bash
cargo clippy --all-targets --all-features -- -D warnings
```

Would you want to add a clippy check to the CI steps? And maybe documentation tests as well?

And I'm not sure if travis CI supports caching, but I've been able to get a huge speed-up in CI by using caching in GitHub Actions.

**Comment by guillaume-be** [comment]
@epwalsh thank you very much for the PR!

I did notice the clippy warnings, I am working on fixing them and they will be addressed in the next release. I will then add a clippy check to the CI. The documentation tests are already run and check as part of the CI.

Regarding caching, I had it enabled at the beginning for this project, but the time required to compress and decompress the large cache was actually longer than downloading the files again.

**Comment by epwalsh** [comment]
Oh right. By “large files” are you talking about model weights / checkpoint files?

I just meant caching the cargo build artifacts. That can make a big difference especially for Clippy, which takes a while to run from scratch every time

**Pull request merged**